Documentation/driver-api/ancillary_bus.rst

```diff
@@ -0,0 +1,215 @@
+Splitting the functionality into smaller orthogonal devices would make it
+easier to manage data, power management and domain-specific interaction with
+the hardware. A key requirement for such a split is that there is no dependency
+on a physical bus, device, register accesses or regmap support. These
+individual devices split from the core cannot live on the platform bus as they
+are not physical devices that are controlled by DT/ACPI. The same argument
+applies for not using MFD in this scenario as MFD relies on individual function
+devices being physical devices that are DT enumerated.
+
+An example for this kind of requirement is the audio subsystem where a single
+IP may handle multiple entities such as HDMI, Soundwire, local devices such as
+mics/speakers etc. The split for the core's functionality can be arbitrary or
+be defined by the DSP firmware topology and include hooks for test/debug. This
+allows for the audio core device to be minimal and focused on hardware-specific
+control and communication.
+
+The ancillary bus is intended to be minimal, generic and avoid domain-specific
+assumptions. Each ancillary_device represents a part of its parent
+functionality. The generic behavior can be extended and specialized as needed
+by encapsulating an ancillary_device within other domain-specific structures and
+the use of .ops callbacks. Devices on the ancillary bus do not share any
+structures and the use of a communication channel with the parent is
+domain-specific.
+
+When Should the Ancillary Bus Be Used
+=====================================
+
+One example could be a multi-port PCI network interface card whose driver will
+allocate and register an ancillary_device for each physical function on the NIC.
+An RDMA driver registers an ancillary_driver that will be matched with and
+probed for each of these ancillary_devices.
+
+Another usage case is for a PCI device whose functionality is complex enough
+that it warrants dividing it into multiple driver modules.  The ancillary_device
+and ancillary_driver matching and probing would allow the main driver to expose
+a shared memory object(s) with the peer modules to allow them to perform their
+desired function.
+
+The ancillary bus is to be used when a driver and one or more kernel modules,
+who share a common header file with the driver, need a mechanism to connect and
+provide access to a shared object allocated by the ancillary_device's
+registering driver.  The registering driver for the ancillary_device(s) and the
+kernel module(s) registering ancillary_drivers can be from the same subsystem,
+or from multiple subsystems.
+
+The emphasis here is on a common generic interface that keeps subsystem
+customization out of the bus infrastructure.
+
+When Should the Ancillary Bus Not Be Used
+=========================================
+
+If there is a need for a more specifically customized bus that provides
+functionality beyond just the matching up of client devices and client drivers,
```

Comment:
client --> ancillary? This is the only place you use client and it looks odd.

Reply:
changed